lib/publiccloud/provider.pm

```diff
@@ -235,14 +236,20 @@ sub get_image_id {
     my ($self, $img_url) = @_;
     my $predefined_id = get_var('PUBLIC_CLOUD_IMAGE_ID');
     return $predefined_id if ($predefined_id);
-    $img_url //= get_required_var('PUBLIC_CLOUD_IMAGE_LOCATION');
-    my ($img_name) = $img_url =~ /([^\/]+)$/;
-    $self->{image_cache} //= {};
-    return $self->{image_cache}->{$img_name} if ($self->{image_cache}->{$img_name});
-    my $image_id = $self->find_img($img_name);
-    die("Image $img_name is not available in the cloud provider") unless ($image_id);
-    $self->{image_cache}->{$img_name} = $image_id;
-    return $image_id;
+    $img_url //= get_var('PUBLIC_CLOUD_IMAGE_LOCATION');
+    if ($img_url) {
+        my ($img_name) = $img_url =~ /([^\/]+)$/;
+        $self->{image_cache} //= {};
+        return $self->{image_cache}->{$img_name} if ($self->{image_cache}->{$img_name});
+        my $image_id = $self->find_img($img_name);
+        die("Image $img_name is not available in the cloud provider") unless ($image_id);
+        $self->{image_cache}->{$img_name} = $image_id;
+        return $image_id;
+    } else {
+        # QEM Azure runs are based on OFFER and SKU variables and do not set PUBLIC_CLOUD_IMAGE_ID or PUBLIC_CLOUD_IMAGE_LOCATION
+        return "" if (is_azure && get_var('PUBLIC_CLOUD_AZURE_OFFER') && get_var('PUBLIC_CLOUD_AZURE_SKU'));
```

Comment:
My inital idea was, to collect all "provider" independent stuff in `publiccloud::provider`. If some provider is doing something special, I would overwrite that specific function in the corresponding provider, like `publiccloud::azure`

Reply:
Of course, my bad :man_facepalming: Moving it to `azure.pm`.